File: es/codes/javascript/chapter_backtracking/preorder_traversal_ii_compact.js
/**
 * File: preorder_traversal_ii_compact.js
 * Created Time: 2023-05-09
 * Author: Justin (xiefahit@gmail.com)
 */

const { arrToTree } = require('../modules/TreeNode');
const { printTree } = require('../modules/PrintUtil');

/* 前序遍历：例题二 */
function preOrder(root, path, res) {
    if (root === null) {
        return;
    }
    // 尝试
    path.push(root);
    if (root.val === 7) {
        // 记录解
        res.push([...path]);
    }
    preOrder(root.left, path, res);
    preOrder(root.right, path, res);
    // 回退
    path.pop();
}

// Driver Code
const root = arrToTree([1, 7, 3, 4, 5, 6, 7]);
console.log('\n初始化二叉树');
printTree(root);

// 前序遍历
const path = [];
const res = [];
preOrder(root, path, res);

console.log('\n输出所有根节点到节点 7 的路径');
res.forEach((path) => {
    console.log(path.map((node) => node.val));
});

File: en/codes/ruby/chapter_backtracking/preorder_traversal_ii_compact.rb
=begin
File: preorder_traversal_ii_compact.rb
Created Time: 2024-05-22
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/tree_node'
require_relative '../utils/print_util'

### Pre-order traversal: example 2 ###
def pre_order(root)
  return unless root

  # Attempt
  $path << root

  # Record solution
  $res << $path.dup if root.val == 7

  pre_order(root.left)
  pre_order(root.right)

  # Backtrack
  $path.pop
end

### Driver Code ###
if __FILE__ == $0
  root = arr_to_tree([1, 7, 3, 4, 5, 6, 7])
  puts "\nInitialize binary tree"
  print_tree(root)

  # Preorder traversal
  $path, $res = [], []
  pre_order(root)

  puts "\nOutput all paths from root node to node 7"
  for path in $res
    p path.map { |node| node.val }
  end
end
